src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -1088,6 +1064,40 @@ public int OnAfterCloseSolution(object pUnkReserved)
         }
         #endregion
 
+        #region IVsSolutionLoadEvents
+        int IVsSolutionLoadEvents.OnBeforeOpenSolution(string pszSolutionFilename)
+        {
+            SolutionDirectory = Path.GetDirectoryName(pszSolutionFilename);
```

Comment:
May be there is a valid reason, but I am not sure why we are overriding this value again in following code in `InitializeAsync` method 
https://github.com/NuGet/NuGet.Client/pull/4164/files#diff-bed53a8b2e737e49ff6141c4279e525dfa2d100eabf6510ef6e0a1746de10505R189

``` cs
            int hr = _vsSolution.GetSolutionInfo(out string soltuionDirectory, out string soltuionFile, out string userOptsFile);
            if (hr == VSConstants.S_OK && !string.IsNullOrEmpty(soltuionDirectory))
            {
                IsSolutionOpen = true;
                SolutionDirectory = soltuionDirectory.TrimEnd('\\');
            }
```

Reply:
VSSolutionManager isn't always initialized before solution load, only when NuGet is needed. Sometimes it is, like when PMC is open on VS open, but otherwise in my debugging `InitializeAsync` is only called once a .NET project is loaded (by which time the solution is already at least partially open).